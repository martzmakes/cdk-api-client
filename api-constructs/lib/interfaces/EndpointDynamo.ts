/**
 * Interface for endpoint definition
 */
export interface EndpointDynamo {
  action?: string;
  defaultLimit?: number;
  description?: string;
  expressionAttributeNames?: Record<string, any>;
  expressionAttributeValues?: Record<string, any>;
  filterExpression?: string;
  indexName?: string;
  input?: any;
  keyConditionExpression?: string;
  method: string;
  output?: any;
  path: string;
  pk?: string;
  requestTemplateOverride?: string;
  responseKey?: string;
  responseTemplateOverride?: string;
  responseTemplateDisableItemNotFoundHandler?: boolean;
  sk?: string;
  tableName: string;
}
