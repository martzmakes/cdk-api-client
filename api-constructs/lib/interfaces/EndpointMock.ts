/**
 * Interface for endpoint mock information
 */
export interface EndpointMock {
  name: string;
  paramType: string;
  returnType: string;
}