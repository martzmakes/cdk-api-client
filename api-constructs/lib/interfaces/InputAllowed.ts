import { HttpMethod } from "./HttpMethod";

export type InputAllowed<M extends HttpMethod, InputType = any> = M extends "GET" | "DELETE"
  ? undefined
  : InputType;