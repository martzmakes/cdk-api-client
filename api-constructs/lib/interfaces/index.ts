export * from "./ApiClientDefinition";
export * from "./ApiEndpoint";
export * from "./Endpoint";
export * from "./EndpointMock";
export * from "./HttpMethod";
export * from "./InputAllowed";
export * from "./TypeImport";