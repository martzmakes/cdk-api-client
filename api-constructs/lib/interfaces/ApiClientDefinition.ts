import { ApiEndpoint } from "./ApiEndpoint";

export type ApiClientDefinition<
  Endpoints extends Record<string, ApiEndpoint<any, any, any>>,
> = {
  [K in keyof Endpoints]: Endpoints[K];
};