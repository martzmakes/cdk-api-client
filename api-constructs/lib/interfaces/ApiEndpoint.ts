import { LambdaProps } from "@martzmakes/constructs/cdk/interfaces/LambdaProps";
import { HttpMethod } from "./HttpMethod";
import { InputAllowed } from "./InputAllowed";

export type ApiEndpoint<
  M extends HttpMethod,
  Input = InputAllowed<M>,
  Output = any,
  Resources = any,
> = {
  description?: string;
  input?: Input;
  method: M;
  output?: Output;
  path: string;
} & (
  | {
      entry: string;
      queue?: boolean;
      lambdaGenerator: (args: Resources) => Omit<LambdaProps, "entry" | "name" | "queue" | "description">;
    }
  // | {
  //     dynamoGenerator: (args: Resources) => Omit<EndpointDynamo, "method" | "path" | "requestTemplate" | "responseTemplate">;
  //     requestTemplate?: string;
  //     responseTemplate: string;
  //   }
);