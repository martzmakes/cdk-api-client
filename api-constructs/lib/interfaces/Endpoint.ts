import { ApiEndpoint } from "./ApiEndpoint";
import { HttpMethod } from "./HttpMethod";

/**
 * Type alias for endpoint definitions
 * 
 * An endpoint can be implemented in two ways:
 * 1. Lambda-based: uses a Lambda function for implementation (entry + lambdaGenerator)
 * 2. DynamoDB-based: uses direct DynamoDB integration (dynamoGenerator + requestTemplate/responseTemplate)
 *    For DynamoDB endpoints, VTL templates will be automatically generated based on input/output types
 */
export type Endpoint = ApiEndpoint<HttpMethod>;