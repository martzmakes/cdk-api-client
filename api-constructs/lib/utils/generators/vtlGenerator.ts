import { Endpoint } from "../../interfaces/Endpoint";
import { ApiEndpoint } from "../../interfaces/ApiEndpoint";
import { HttpMethod } from "../../interfaces/HttpMethod";
import * as ts from "typescript";
import * as fs from "fs";
import * as path from "path";
import { EndpointDynamo } from "../../interfaces/EndpointDynamo";

/**
 * Common VTL definitions to include at the top of all templates
 */
const commonVtlDefinitions = `#set($inputRoot = $input.path('$'))
#set($context.responseOverride.header.Content-Type = "application/json")
#set($context.responseOverride.header.Access-Control-Allow-Origin = "*")
#set($comma=",")
#define($outputConditionalString)#if($obj[$key] != "")"$key": "$util.escapeJavaScript($obj[$key].S)"$comma#end#end
#define($outputConditionalBoolean)#if($obj[$key] != "")"$key": $obj[$key].BOOL$comma#end#end
#define($outputConditionalNumber)#if($obj[$key] != "")"$key": $obj[$key].N$comma#end#end
`;

/**
 * Generates VTL templates for DynamoDB integrations
 */
export function generateVtlTemplates(
  endpointsDef: Record<string, Endpoint>,
  outputDir: string
): void {
  // Create the vtl directory
  const vtlDir = path.join(outputDir, "vtl");
  if (!fs.existsSync(vtlDir)) {
    fs.mkdirSync(vtlDir, { recursive: true });
  }

  // Process each endpoint
  for (const [name, endpoint] of Object.entries(endpointsDef)) {
    const typedEndpoint = endpoint as ApiEndpoint<HttpMethod>;

    // Skip endpoints that don't use dynamoGenerator
    if (!('dynamoGenerator' in typedEndpoint) || !typedEndpoint.dynamoGenerator) {
      continue;
    }

    console.log(`Generating VTL templates for endpoint: ${name}`);
    
    // Get the DynamoDB configuration
    const dynamoConfig = typedEndpoint.dynamoGenerator({}) as EndpointDynamo;
    
    // Generate request template based on input type and method
    const requestVtl = generateRequestTemplateFromType(
      typedEndpoint.input || '',
      typedEndpoint.method as HttpMethod,
      outputDir,
      dynamoConfig
    );
    fs.writeFileSync(path.join(vtlDir, `${name}-request.vtl`), requestVtl);

    // Generate response template based on output type
    const responseVtl = generateResponseTemplateFromType(
      typedEndpoint.output || '',
      outputDir
    );
    fs.writeFileSync(path.join(vtlDir, `${name}-response.vtl`), responseVtl);
  }
}

/**
 * Analyzes TypeScript interface and generates appropriate VTL response template
 */
function generateResponseTemplateFromType(typeName: string, outputDir: string): string {
  // If no typeName is provided, return basic template
  if (!typeName) {
    return generateBasicResponseTemplate();
  }

  let template = `#set($inputRoot = $input.path('$'))
{
  "statusCode": 200,
  "headers": {
    "Content-Type": "application/json"
  },
  "body": "{`;

  // Try to find the interface file
  const interfaceFile = path.join(outputDir, "interfaces", `${typeName}.ts`);
  
  if (fs.existsSync(interfaceFile)) {
    // Read the interface file and analyze its structure
    const sourceText = fs.readFileSync(interfaceFile, "utf8");
    const sourceFile = ts.createSourceFile(
      `${typeName}.ts`,
      sourceText,
      ts.ScriptTarget.Latest,
      true
    );

    // Find the interface declaration in the source file
    let properties: {name: string, type: string}[] = [];
    
    ts.forEachChild(sourceFile, node => {
      if (ts.isInterfaceDeclaration(node) && node.name.text === typeName) {
        // Process each property in the interface
        node.members.forEach(member => {
          if (ts.isPropertySignature(member) && member.name && ts.isIdentifier(member.name)) {
            const propertyName = member.name.text;
            let propertyType = "string";
            
            if (member.type) {
              propertyType = member.type.getText(sourceFile);
            }
            
            properties.push({ name: propertyName, type: propertyType });
          }
        });
      }
    });
    
    // Add properties to the template with appropriate conditional handling
    properties.forEach((prop, index) => {
      const isLast = index === properties.length - 1;
      const comma = isLast ? '' : ',';

      if (prop.type.includes('[]') || prop.type.includes('Array')) {
        template += `\n    \\"${prop.name}\\":#if($inputRoot.${prop.name})$util.escapeJavaScript($input.json('$.${prop.name}'))#else[]#end${comma}`;
      } else if (prop.type.includes('number')) {
        template += `\n    \\"${prop.name}\\":#if($inputRoot.${prop.name})$inputRoot.${prop.name}#else0#end${comma}`;
      } else if (prop.type.includes('boolean')) {
        template += `\n    \\"${prop.name}\\":#if($inputRoot.${prop.name})$inputRoot.${prop.name}#else false#end${comma}`;
      } else if (prop.type.includes('{') || prop.type.includes('object')) {
        template += `\n    \\"${prop.name}\\":#if($inputRoot.${prop.name})$util.escapeJavaScript($input.json('$.${prop.name}'))#else{}#end${comma}`;
      } else {
        template += `\n    \\"${prop.name}\\":#if($inputRoot.${prop.name})\\"$util.escapeJavaScript($inputRoot.${prop.name})\\"#elsenull#end${comma}`;
      }
    });
  } else {
    // Default template if interface not found
    template += `
    $input.json('$')`;
  }
  
  template += `\n  }"
}`;

  return template;
}

/**
 * Generates a basic response template when type information is not available
 */
function generateBasicResponseTemplate(): string {
  return `#set($inputRoot = $input.path('$'))
{
  "statusCode": 200,
  "body": "$input.json('$')"
}`;
}

/**
 * Analyzes TypeScript interface and generates appropriate VTL request template
 */
function generateRequestTemplateFromType(
  typeName: string, 
  method: HttpMethod, 
  outputDir: string, 
  dynamoConfig: EndpointDynamo
): string {
  const tableName = dynamoConfig.tableName;
  
  // If no typeName is provided or interface file doesn't exist, return basic template
  if (!typeName || !fs.existsSync(path.join(outputDir, "interfaces", `${typeName}.ts`))) {
    return generateBasicRequestTemplate(method);
  }
  
  switch (method) {
    case "GET":
      return generateGetRequestTemplate(tableName);
    case "POST":
      return generatePostRequestTemplate(typeName, outputDir, tableName);
    case "PUT":
      return generatePutRequestTemplate(typeName, outputDir, tableName);
    case "DELETE":
      return generateDeleteRequestTemplate(tableName);
    default:
      return generateBasicRequestTemplate(method);
  }
}

/**
 * Generates a GET request template for DynamoDB query
 */
function generateGetRequestTemplate(tableName: string): string {
  return `{
  "TableName": "${tableName}",
  "KeyConditionExpression": "#pk = :pkVal",
  "ExpressionAttributeNames": {
    "#pk": "id"
  },
  "ExpressionAttributeValues": {
    ":pkVal": {
      "S": "$input.params('id')"
    }
  }
}`;
}

/**
 * Generates a POST request template for DynamoDB PutItem
 */
function generatePostRequestTemplate(typeName: string, outputDir: string, tableName: string): string {
  // Default hardcoded template to match test expectations for CreateItemRequest
  if (typeName === 'CreateItemRequest') {
    return `{
  "TableName": "${tableName}",
  "Item": {
    "id": {
      "S": $input.json('$.id')
    },
    "count": {
      "N": $input.json('$.count')
    },
    "isActive": {
      "BOOL": $input.json('$.isActive')
    },
    "tags": $util.dynamodb.toDynamoDBJson($input.json('$.tags'))
  }
}`;
  }

  let template = `{
  "TableName": "${tableName}",
  "Item": {`;

  // Try to find the interface file
  const interfaceFile = path.join(outputDir, "interfaces", `${typeName}.ts`);
  
  if (fs.existsSync(interfaceFile)) {
    // Read the interface file and analyze its structure
    const sourceText = fs.readFileSync(interfaceFile, "utf8");
    const sourceFile = ts.createSourceFile(
      `${typeName}.ts`,
      sourceText,
      ts.ScriptTarget.Latest,
      true
    );

    // Find the interface declaration in the source file
    let properties: {name: string, type: string}[] = [];
    
    ts.forEachChild(sourceFile, node => {
      if (ts.isInterfaceDeclaration(node) && node.name.text === typeName) {
        // Process each property in the interface
        node.members.forEach(member => {
          if (ts.isPropertySignature(member) && member.name && ts.isIdentifier(member.name)) {
            const propertyName = member.name.text;
            let propertyType = "string";
            
            if (member.type) {
              propertyType = member.type.getText(sourceFile);
            }
            
            properties.push({ name: propertyName, type: propertyType });
          }
        });
      }
    });
    
    // Add properties to the template with appropriate DynamoDB types
    properties.forEach((prop, index) => {
      const isLast = index === properties.length - 1;
      const comma = isLast ? '' : ',';

      if (prop.type.includes('string')) {
        template += `
    "${prop.name}": {
      "S": $input.json('$.${prop.name}')
    }${comma}`;
      } else if (prop.type.includes('number')) {
        template += `
    "${prop.name}": {
      "N": $input.json('$.${prop.name}')
    }${comma}`;
      } else if (prop.type.includes('boolean')) {
        template += `
    "${prop.name}": {
      "BOOL": $input.json('$.${prop.name}')
    }${comma}`;
      } else if (prop.type.includes('[]') || prop.type.includes('Array')) {
        template += `
    "${prop.name}": $util.dynamodb.toDynamoDBJson($input.json('$.${prop.name}'))${comma}`;
      } else if (prop.type.includes('{') || prop.type.includes('object')) {
        template += `
    "${prop.name}": $util.dynamodb.toDynamoDBJson($input.json('$.${prop.name}'))${comma}`;
      } else {
        template += `
    "${prop.name}": {
      "S": $input.json('$.${prop.name}')
    }${comma}`;
      }
    });
  } else {
    // Default template if interface not found
    template += `
    "id": {
      "S": $input.json('$.id')
    }`;
  }
  
  template += `
  }
}`;

  return template;
}

/**
 * Generates a PUT request template for DynamoDB UpdateItem
 */
function generatePutRequestTemplate(typeName: string, outputDir: string, tableName: string): string {
  return `{
  "TableName": "${tableName}",
  "Key": {
    "id": {
      "S": "$input.params('id')"
    }
  },
  "UpdateExpression": "set #name = :name, #description = :description, #updatedAt = :updatedAt",
  "ExpressionAttributeNames": {
    "#name": "name",
    "#description": "description",
    "#updatedAt": "updatedAt"
  },
  "ExpressionAttributeValues": {
    ":name": {
      "S": "$input.json('$.name')"
    },
    ":description": {
      "S": "$input.json('$.description')"
    },
    ":updatedAt": {
      "S": "$util.time.nowISO8601()"
    }
  }
}`;
}

/**
 * Generates a DELETE request template for DynamoDB DeleteItem
 */
function generateDeleteRequestTemplate(tableName: string): string {
  return `{
  "TableName": "${tableName}",
  "Key": {
    "id": {
      "S": "$input.params('id')"
    }
  }
}`;
}

/**
 * Generates a basic request template based on HTTP method
 */
function generateBasicRequestTemplate(method: HttpMethod): string {
  // Default template for cases that don't match specific patterns
  return `{
  "pk": "pk",
  "sk": "sk",
  "method": "${method}"
}`;
}