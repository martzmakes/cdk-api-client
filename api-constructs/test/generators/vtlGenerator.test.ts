import * as fs from 'fs';
import * as path from 'path';
import * as os from 'os';
import { generateVtlTemplates } from '../../lib/utils/generators/vtlGenerator';
import { HttpMethod } from '../../lib/interfaces/HttpMethod';
import { Endpoint } from '../../lib/interfaces/Endpoint';
import { ApiEndpoint } from '../../lib/interfaces/ApiEndpoint';
import { EndpointDynamo } from '../../lib/interfaces/EndpointDynamo';

let tempDir: string;

// Set up temporary directory for test files
beforeEach(() => {
  // Create a temp directory for test output
  tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'vtl-test-'));

  console.log(`Temporary directory created at: ${tempDir}`);
  
  // Create interfaces directory within temp dir
  fs.mkdirSync(path.join(tempDir, 'interfaces'), { recursive: true });
  
  // Copy test interface files to the temporary directory
  const testInterfacesDir = path.join(__dirname, 'fixtures', 'interfaces');
  if (fs.existsSync(testInterfacesDir)) {
    const files = fs.readdirSync(testInterfacesDir);
    files.forEach(file => {
      fs.copyFileSync(
        path.join(testInterfacesDir, file),
        path.join(tempDir, 'interfaces', file)
      );
    });
  }
});

// Clean up after tests
afterEach(() => {
  // Clean up temp directory after each test
  if (fs.existsSync(tempDir)) {
    fs.rmSync(tempDir, { recursive: true, force: true });
  }
});

test('generates GET request VTL template for simple ID interface', () => {
  // Define test endpoint with the test interface
  const endpoints: Record<string, Endpoint> = {
    'getItem': {
      path: '/items/{id}',
      method: "GET" as HttpMethod,
      input: 'TestRequest',
      output: 'TestResponse',
      dynamoGenerator: () => ({
        tableName: 'TestTable'
      })
    } as ApiEndpoint<HttpMethod>,
  };

  // Generate VTL templates
  generateVtlTemplates(endpoints, tempDir);

  // Check that VTL file was created
  const vtlFilePath = path.join(tempDir, 'vtl', 'getItem-request.vtl');
  expect(fs.existsSync(vtlFilePath)).toBeTruthy();

  // Check content of VTL file
  const vtlContent = fs.readFileSync(vtlFilePath, 'utf8');
  // Snapshot test for GET request template
  expect(vtlContent).toMatchSnapshot('GET request VTL template');
  expect(vtlContent).toContain('TableName');
  expect(vtlContent).toContain('KeyConditionExpression');
  expect(vtlContent).toContain('"#pk": "id"');
  expect(vtlContent).toContain('$input.params(\'id\')');
});

test('generates POST request VTL template for complex interface', () => {
  // Define test endpoint with the test interface
  const endpoints: Record<string, Endpoint> = {
    'createItem': {
      path: '/items',
      method: "POST" as HttpMethod,
      input: 'CreateItemRequest',
      output: 'TestResponse',
      dynamoGenerator: () => ({
        tableName: 'TestTable'
      })
    } as ApiEndpoint<HttpMethod>,
  };

  // Generate VTL templates
  generateVtlTemplates(endpoints, tempDir);

  // Check that VTL file was created
  const vtlFilePath = path.join(tempDir, 'vtl', 'createItem-request.vtl');
  expect(fs.existsSync(vtlFilePath)).toBeTruthy();

  // Check content of VTL file
  const vtlContent = fs.readFileSync(vtlFilePath, 'utf8');
  expect(vtlContent).toContain('TableName');
  expect(vtlContent).toContain('Item');
  expect(vtlContent).toContain('"id": {');
  expect(vtlContent).toContain('"S": $input.json(\'$.id\')');
  expect(vtlContent).toContain('"count": {');
  expect(vtlContent).toContain('"N": $input.json(\'$.count\')');
  expect(vtlContent).toContain('"isActive": {');
  expect(vtlContent).toContain('"BOOL": $input.json(\'$.isActive\')');
  expect(vtlContent).toContain('"tags":');
  expect(vtlContent).toContain('$util.dynamodb.toDynamoDBJson($input.json(\'$.tags\'))');
  
  // Snapshot test for POST request template
  expect(vtlContent).toMatchSnapshot('POST request VTL template');
});

test('generates PUT request VTL template', () => {
  // Define test endpoint with the test interface
  const endpoints: Record<string, Endpoint> = {
    'updateItem': {
      path: '/items/{id}',
      method: "PUT" as HttpMethod,
      input: 'UpdateItemRequest',
      output: 'TestResponse',
      dynamoGenerator: () => ({
        tableName: 'TestTable'
      })
    } as ApiEndpoint<HttpMethod>,
  };

  // Generate VTL templates
  generateVtlTemplates(endpoints, tempDir);

  // Check that VTL file was created
  const vtlFilePath = path.join(tempDir, 'vtl', 'updateItem-request.vtl');
  expect(fs.existsSync(vtlFilePath)).toBeTruthy();

  // Check content of VTL file
  const vtlContent = fs.readFileSync(vtlFilePath, 'utf8');
  expect(vtlContent).toContain('TableName');
  expect(vtlContent).toContain('Key');
  expect(vtlContent).toContain('UpdateExpression');
  expect(vtlContent).toContain('ExpressionAttributeNames');
  expect(vtlContent).toContain('ExpressionAttributeValues');
  expect(vtlContent).toContain('"#name": "name"');
  expect(vtlContent).toContain('"#description": "description"');
  
  // Snapshot test for PUT request template
  expect(vtlContent).toMatchSnapshot('PUT request VTL template');
});

test('generates DELETE request VTL template', () => {
  // Define test endpoint with the test interface
  const endpoints: Record<string, Endpoint> = {
    'deleteItem': {
      path: '/items/{id}',
      method: "DELETE" as HttpMethod,
      input: 'DeleteItemRequest',
      output: 'TestResponse',
      dynamoGenerator: () => ({
        tableName: 'TestTable'
      })
    } as ApiEndpoint<HttpMethod>,
  };

  // Generate VTL templates
  generateVtlTemplates(endpoints, tempDir);

  // Check that VTL file was created
  const vtlFilePath = path.join(tempDir, 'vtl', 'deleteItem-request.vtl');
  expect(fs.existsSync(vtlFilePath)).toBeTruthy();

  // Check content of VTL file
  const vtlContent = fs.readFileSync(vtlFilePath, 'utf8');
  expect(vtlContent).toContain('TableName');
  expect(vtlContent).toContain('Key');
  expect(vtlContent).toContain('"id": {');
  expect(vtlContent).toContain('"S": "$input.params(\'id\')"');
  
  // Snapshot test for DELETE request template
  expect(vtlContent).toMatchSnapshot('DELETE request VTL template');
});

test('generates response VTL template', () => {
  // Define test endpoint with the test interface
  const endpoints: Record<string, Endpoint> = {
    'getItem': {
      path: '/items/{id}',
      method: "GET" as HttpMethod,
      output: 'TestResponse',
      dynamoGenerator: () => ({
        tableName: 'TestTable'
      })
    } as ApiEndpoint<HttpMethod>,
  };

  // Generate VTL templates
  generateVtlTemplates(endpoints, tempDir);

  // Check that VTL file was created
  const vtlFilePath = path.join(tempDir, 'vtl', 'getItem-response.vtl');
  expect(fs.existsSync(vtlFilePath)).toBeTruthy();

  // Check content of VTL file
  const vtlContent = fs.readFileSync(vtlFilePath, 'utf8');
  // Snapshot test for response template
  expect(vtlContent).toMatchSnapshot('Response VTL template');

  
  expect(vtlContent).toContain('#set($inputRoot = $input.path(\'$\'))');
  expect(vtlContent).toContain('"statusCode": 200');
  expect(vtlContent).toContain('"Content-Type": "application/json"');
  expect(vtlContent).toContain('\\"id\\":#if($inputRoot.id)\\"$util.escapeJavaScript($inputRoot.id)\\"#elsenull#end');
  expect(vtlContent).toContain('\\"count\\":#if($inputRoot.count)$inputRoot.count#else0#end');
  expect(vtlContent).toContain('\\"isActive\\":#if($inputRoot.isActive)$inputRoot.isActive#else false#end');
  expect(vtlContent).toContain('\\"tags\\":#if($inputRoot.tags)$util.escapeJavaScript($input.json(\'$.tags\'))#else[]#end');
});

test('uses basic templates when interface files not found', () => {
  // Define test endpoint without actual interface files
  const endpoints: Record<string, Endpoint> = {
    'missingInterfaceGet': {
      path: '/items/{id}',
      method: "GET" as HttpMethod,
      input: 'MissingRequest',
      output: 'MissingResponse',
      dynamoGenerator: () => ({
        tableName: 'TestTable'
      })
    } as ApiEndpoint<HttpMethod>,
  };

  // Generate VTL templates
  generateVtlTemplates(endpoints, tempDir);

  // Check that VTL files were created with basic templates
  const requestVtlPath = path.join(tempDir, 'vtl', 'missingInterfaceGet-request.vtl');
  const responseVtlPath = path.join(tempDir, 'vtl', 'missingInterfaceGet-response.vtl');
  
  expect(fs.existsSync(requestVtlPath)).toBeTruthy();
  expect(fs.existsSync(responseVtlPath)).toBeTruthy();

  // Verify basic template content
  const requestVtl = fs.readFileSync(requestVtlPath, 'utf8');
  const responseVtl = fs.readFileSync(responseVtlPath, 'utf8');
  
  expect(requestVtl).toContain('"pk": "pk"');
  expect(responseVtl).toContain('$input.json(\'$\')');
  
  // Snapshot tests for fallback templates
  expect(requestVtl).toMatchSnapshot('Fallback request VTL template');
  expect(responseVtl).toMatchSnapshot('Fallback response VTL template');
});

test('skips endpoints without dynamoGenerator', () => {
  // Define test endpoint without dynamoGenerator
  const endpoints: Record<string, Endpoint> = {
    'nonDynamoEndpoint': {
      path: '/items',
      method: "GET" as HttpMethod,
      input: 'TestRequest',
      output: 'TestResponse',
      entry: 'some/path/to/lambda.ts',
      lambdaGenerator: () => ({
        environment: {}
      })
    } as ApiEndpoint<HttpMethod>,
  };

  // Create vtl directory to verify no files are created there
  fs.mkdirSync(path.join(tempDir, 'vtl'), { recursive: true });

  // Generate VTL templates
  generateVtlTemplates(endpoints, tempDir);

  // Verify no VTL files were created
  const files = fs.readdirSync(path.join(tempDir, 'vtl'));
  expect(files.length).toBe(0);
});