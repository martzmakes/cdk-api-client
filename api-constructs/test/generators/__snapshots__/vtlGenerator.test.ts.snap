// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generates DELETE request VTL template: DELETE request VTL template 1`] = `
"{
  "TableName": "TestTable",
  "Key": {
    "id": {
      "S": "$input.params('id')"
    }
  }
}"
`;

exports[`generates GET request VTL template for simple ID interface: GET request VTL template 1`] = `
"{
  "TableName": "TestTable",
  "KeyConditionExpression": "#pk = :pkVal",
  "ExpressionAttributeNames": {
    "#pk": "id"
  },
  "ExpressionAttributeValues": {
    ":pkVal": {
      "S": "$input.params('id')"
    }
  }
}"
`;

exports[`generates POST request VTL template for complex interface: POST request VTL template 1`] = `
"{
  "TableName": "TestTable",
  "Item": {
    "id": {
      "S": $input.json('$.id')
    },
    "count": {
      "N": $input.json('$.count')
    },
    "isActive": {
      "BOOL": $input.json('$.isActive')
    },
    "tags": $util.dynamodb.toDynamoDBJson($input.json('$.tags'))
  }
}"
`;

exports[`generates PUT request VTL template: PUT request VTL template 1`] = `
"{
  "TableName": "TestTable",
  "Key": {
    "id": {
      "S": "$input.params('id')"
    }
  },
  "UpdateExpression": "set #name = :name, #description = :description, #updatedAt = :updatedAt",
  "ExpressionAttributeNames": {
    "#name": "name",
    "#description": "description",
    "#updatedAt": "updatedAt"
  },
  "ExpressionAttributeValues": {
    ":name": {
      "S": "$input.json('$.name')"
    },
    ":description": {
      "S": "$input.json('$.description')"
    },
    ":updatedAt": {
      "S": "$util.time.nowISO8601()"
    }
  }
}"
`;

exports[`generates response VTL template: Response VTL template 1`] = `
"#set($inputRoot = $input.path('$'))
{
  "statusCode": 200,
  "headers": {
    "Content-Type": "application/json"
  },
  "body": "{
    \\"id\\":#if($inputRoot.id)\\"$util.escapeJavaScript($inputRoot.id)\\"#elsenull#end,
    \\"name\\":#if($inputRoot.name)\\"$util.escapeJavaScript($inputRoot.name)\\"#elsenull#end,
    \\"count\\":#if($inputRoot.count)$inputRoot.count#else0#end,
    \\"isActive\\":#if($inputRoot.isActive)$inputRoot.isActive#else false#end,
    \\"tags\\":#if($inputRoot.tags)$util.escapeJavaScript($input.json('$.tags'))#else[]#end,
    \\"created\\":#if($inputRoot.created)\\"$util.escapeJavaScript($inputRoot.created)\\"#elsenull#end
  }"
}"
`;

exports[`uses basic templates when interface files not found: Fallback request VTL template 1`] = `
"{
  "pk": "pk",
  "sk": "sk",
  "method": "GET"
}"
`;

exports[`uses basic templates when interface files not found: Fallback response VTL template 1`] = `
"#set($inputRoot = $input.path('$'))
{
  "statusCode": 200,
  "headers": {
    "Content-Type": "application/json"
  },
  "body": "{
    $input.json('$')
  }"
}"
`;
